### PR TITLE
Assignment #4

### PR DESCRIPTION
**Pull request by ustya-k**

### CODE REVIEW

**Comment by ustya-k** [comment]
Forgot that assignment 3 would also be here, I'll just update it later on both branches. If you now more efficient way to do this, please, tell.

**Comment by flerdacodeu** [comment]
I think the reason the Week 3 assignment appears in both branches is because you created the week 4 branch when looking at the week 3 branch. If you switched to master first, the new branch would have not included week 3. That said, it is not a big issue. I will ignore the week 3 files in this pull request. Once you merge week 3 into master, the pull request will no longer whose them.

File: week4/Q1.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	for i, row in enumerate(islands_map):
+		for j, tile in enumerate(row):
+			if tile == old_name:
+				islands_map[i][j] = new_name
+	return islands_map
+
+
+def _name_island(islands_map, row, column, islands, island_counter, upper_tile, left_tile):
+	'''
+	Gives name (which island) to the new tile, renames old tiles if needed (if islands which were seen as separate actually happen to be connected).
```
Comment (flerdacodeu) [inline week4/Q1.py:22]: Generally, keep your line at most 80 characters wide, so that when reading them, one does not have to scroll left and right. 80 is arbitrary, but a common limit.

Comment (flerdacodeu) [inline week4/Q1.py:22]: This comment is not clear: I cannot understand by reading this what the function is meant to do.
It would be nice to rephrase it to be clearer and make sure you refer to things by the right name.

For example, you talk about an island name, but I think that refers to the `island_counter`.

Moreover, make sure you use clear names for things.

For example, `island_counter` would seem to contain the number of islands, but you also use
the counter as the name of an island. It is okay to do so, but it would be nicer if you can abstract
that, so that counters and names are not mixed in the code.

File: week4/Q1.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		islands_map: two_dimensional list
+		row: int, row of the tile
+		column: int, column of the tile
+		islands: set, already existing islands
+		island_counter: int, last name given
+		upper_tile: int, value of the tile under which is new one
+		left_tile: int, value of the tile right to which is new one
+
+	Returns:
+		two-dimensional list, set, int
```
Comment (flerdacodeu) [inline week4/Q1.py:34]: Besides the types of the returned elements, describe what they mean: what is in the third elements of the returned tuple? It is an `int`, but what does it represent?

File: week4/Q1.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		elif islands_map[row][column] != left_tile:
+			islands.discard(left_tile)
+			islands_map = _rename_islands(islands_map, left_tile, islands_map[row][column])
+
+	if islands_map[row][column] == 0:
+		island_counter += 1
+		islands_map[row][column] = island_counter
+		islands.add(island_counter)
+
+	return islands_map, islands, island_counter
```
Comment (flerdacodeu) [inline week4/Q1.py:51]: Arrays in Python are passed by reference.

That means that if a function receives an array and modifies it, the caller's array is also modified (there is no copy). Therefore, the function does not need to return the array itself. In fact, this is misleading, because one might assume that the original is left unmodified because one is returned (it might be a copy that has been modified while the original is returned).

For example:
```python
def reset_matrix(matrix, value=0):
    """Resets all the values stored in a two-dimensional matrix.

    Args:
      matrix: a list of lists, the two-dimensional matrix.
      value: the value to set all entries in the matrix to. Default is 0.
    """
    for row in range(len(matrix)):
        for col in range(len(matrix[row])):
            matrix[row][col] = value

matrix = [
        [1, 0, 0],
        [0, 1, 0],
        [0, 0, 1],
]

print('Original:')
print(matrix)
reset_matrix(matrix)
print('Reset:')
print(matrix)
reset_matrix(matrix, value=42)
print('Reset to 42:')
print(matrix)
```
prints:
```python
Original:
[[1, 0, 0], [0, 1, 0], [0, 0, 1]]
Reset:
[[0, 0, 0], [0, 0, 0], [0, 0, 0]]
Reset to 42:
[[42, 42, 42], [42, 42, 42], [42, 42, 42]]
```

This is reasonably common in Python.

For example, `list.sort()` modifies the original list and does not return anything:
```python
>>> l = [2, 3, 1]
>>> l.sort()
>>> l
[1, 2, 3]
```
while `string.upper()` does not modify the original string (because strings are immutable in Python) and therefore has to return the new string.
```python
>>> s = 'AbcD'
>>> s.upper()
'ABCD'
>>> s
'AbcD'
```

One important thing in writing good code is that it behaves in a way that is intuitive and consistent with people's expectations.

File: week4/Q1.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+					left_tile = islands_map[i][j - 1]
+				islands_map, islands, island_counter = _name_island(islands_map, i, j, islands, island_counter, upper_tile, left_tile)
+
+	return len(islands)
+
+
+
+
+def main():
+	n = int(input('Input number of rows: '))
```
Comment (flerdacodeu) [inline week4/Q1.py:86]: For extra credit, you can change this code to be able to read a *textual map*.

Let's define a *textual map* as a string that contains characters representing the islands, where `' '` represents water and `'X'` represents land. The *rows* in the map are separated by newlines.

You should define a function:
```python
def parse_textual_map(string):
  [...]
```
such that:
```python
>>> s = ' X X\nXX  \n  X \n  X ';
>>> print s
 X X
XX
  X
  X
>>> parse_texual_map(s)
[[False, True, False, True], [True, True, False, False], [False, False, True, False], [False, False, True, False]]
```
then the user can simply input the *textual map*. You should stop reading when you have found an empty line, instead of having to ask for the number of rows and columns. You can also provide additional error handling (like not require each row to have spaces at the end and add those for the user).

This code will also be quite helpful for testing, as you can then create a few different maps easily.

File: week4/Q1_test.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import Q1
+import unittest
+
+
+class IslandCountTest(unittest.TestCase):		
+
+	def testBaseCase(self):
```
Comment (flerdacodeu) [inline week4/Q1_test.py:7]: I think there are a few other interesting cases.
What if the map has no land at all?
What if the map has only land?
You can also have specific tests for interesting cases like:
```
X   
 XXX
```
where you want to verify that the diagonal connection does not count.

File: week4/Q1.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+	return islands_map, islands, island_counter
+
+
+def count_islands(tiles_map, n_rows, n_columns):
+	'''
+	Counts number of islands on the given map.
+
+	Args:
+		tiles_map: two-dimensional list of bools
+		n_rows: int
```
Comment (flerdacodeu) [inline week4/Q1.py:60]: While the assignment explicitly mentioned having number of rows and columns as arguments, it is not strictly necessary in your implementation. Because the size of the map is given by `tiles_map`.

This is an example of how the requirements have changed from the original problem formulation to when you actually write the solution (there are solutions that do require the number of rows and columns, but in your case, the code does not require them).

File: week4/Q1.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+	Returns:
+		int
+	'''
+	islands_map = [[0 for i in range(n_columns)] for row in range(n_rows)]
+	island_counter = 0
+	islands = set()
+	for i, row in enumerate(tiles_map):
+		for j, tile in enumerate(row):
+			if tile:
+				upper_tile = None
```
Comment (flerdacodeu) [inline week4/Q1.py:72]: I think it would be helpful to make it clearer that `0` means a few different things in this code (in `upper_tile`, `left_tile`, and `islands_map`).

You are using `None` if the tile above or to the left does not exist. You are using `0` if the tile above or to the left is actually water. You are also using `0` in the islands_map both for the tiles that are water and those that are land but are not yet assigned. I think the use of `None` is actually unnecessary (you could use `0` instead) or you should use it everywhere to mean the same thing (e.g., make all water tiles be `None` and then you can work with the assumption that there is an infinite sea around the map and therefore `upper_tile` is `None` when the tile above is in the infinite sea around the map).

**Review by flerdacodeu — CHANGES_REQUESTED** [review]

**Comment by flerdacodeu** [comment]
Nicely done Ustinia!
The algorithm is spot on.
I have some suggestions for you to make it clearer and adding some test scenarios.
